source/extensions/transport_sockets/alts/tsi_socket.cc

```diff
@@ -259,6 +259,57 @@ Network::IoResult TsiSocket::doRead(Buffer::Instance& buffer) {
   return repeatReadAndUnprotect(buffer, result);
 }
 
+Network::IoResult TsiSocket::repeatProtectAndWrite(Buffer::Instance& buffer, bool end_stream) {
+  uint64_t total_bytes_written = 0;
+  Network::IoResult result = {Network::PostIoAction::KeepOpen, 0, false};
+
+  while (true) {
+    uint64_t bytes_to_drain_this_iteration =
+        prev_bytes_to_drain_ > 0 ? prev_bytes_to_drain_
+                                 : std::min(buffer.length(), actual_frame_size_to_use_);
+    // Consumed all data. Exit.
+    if (bytes_to_drain_this_iteration == 0) {
+      break;
+    }
+    // Short write did not occur previously.
+    if (raw_write_buffer_.length() == 0) {
+      ASSERT(frame_protector_);
+      ASSERT(prev_bytes_to_drain_ == 0);
+
+      // Store data to be protected this iteration in a separate buffer.
+      Buffer::OwnedImpl bytes_to_protect_this_iteration;
+      bytes_to_protect_this_iteration.add(buffer.linearize(bytes_to_drain_this_iteration),
+                                          bytes_to_drain_this_iteration);
+
+      // Do protect.
+      ENVOY_CONN_LOG(debug, "TSI: protecting buffer size: {}", callbacks_->connection(),
+                     bytes_to_protect_this_iteration.length());
+      tsi_result status =
+          frame_protector_->protect(bytes_to_protect_this_iteration, raw_write_buffer_);
```

Comment:
Consider changing the protect method to accept a string_view with the input bytes to avoid copying into bytes_to_protect_this_iteration just to copy again inside protect as it populates the grpc slice.

I think that ExternallyManagedSlice could be used instead of UnmanagedMemorySlice to save another copy during protect, but we can sort that out later.

Reply:
Good point. Yeah, it does not make sense to make bytes copy twice.